Preliminary changes for multiple Compaction Groups

raphaelsc opened the pull request:
What's contained in this series:
- Refactored compaction tests (and utilities) for integration with multiple groups
    - The idea is to write a new class of tests that will stress multiple groups, whereas the existing ones will still stress a single group.
- Fixed a problem when cloning compound sstable set (cannot be triggered today so I didn't open a GH issue)
- Many changes in replica::table for allowing integration with multiple groups

Next:
- Introduce for_each_compaction_group() for iterating over groups wherever needed.
- Use for_each_compaction_group() in replica::table operations spanning all groups (API, readers, etc).
- Decouple backlog tracker from compaction strategy, to allow for backlog isolation across groups
- Introduce static option for defining number of compaction groups and implement function to map a token to its respective group.
- Testing infrastructure for multiple compaction groups (helpful when testing the dynamic behavior: i.e. merging / splitting).

denesb commented on replica/table.cc:
```diff
@@ -1222,8 +1222,7 @@ future<std::unordered_set<sstring>> table::get_sstables_by_partition_key(const s
-    // main sstables is enough for the outside world. sstables in other set like maintenance is not needed even for expiration purposes in compaction
-    return *_compaction_group->main_sstables();
+    return *_sstables;
 }
```
Are there any current users of this method who might be suprised by the change?

raphaelsc replied:
Actually, I was worried about current users of this method who might be getting incorrect behavior due to maintanenance set being left out :-) Compaction layer was one of the users and it's now interacting with table through table_state which exposes main sstables only from the compaction group.

raphaelsc replied:
The change to limit get_sstable_set() to sstables in main set was done after we introduced compound set for off-strategy compaction. Maybe it wasn't the right decision as one could incorrectly assume get_sstable_set() will provide its caller with all sstables (including the ones in maintenance set). But now with table_state() we no longer need to limit this method to main_sstables, but rather return the compound set which spans all the sstables sets in a table.

avikivity replied:
So what's the answer? Will this change cause a bug, or is it fixing a bug, or are all current callers unaffected?

raphaelsc replied:
Turns out it fixes a problem in table::get_sstables_by_partition_key(), which won't see the sstables in maintenance set until reshape completes. That function is exposed via API only.

Will extract the fix into a separate series, and add an unit test.

I also would like to understand the impact of the change in view_builder::initialize_reader_at_current_token().
@nyh, please help. We do register maintenance sstables (stored in maintenance set) for view update using view_update_generator::register_staging_sstable(), so they're consumed by view_update_generator loop. But view_builder::initialize_reader_at_current_token() uses table::get_sstable_set() to create a local reader for a given current token. What's the impact if that reader misses sstables that are temporarily stored in the table's maintenance set (but which were already registered in view_update_generator as mentioned above)?

denesb reviewed (approved):
Looks good, just one question.
Maybe we should rename `column_family_for_tests` to `table_for_tests`. Can be a follow-up.

raphaelsc commented:
> Looks good, just one question. Maybe we should rename `column_family_for_tests` to `table_for_tests`. Can be a follow-up.

Agree 100%. column_family is deprecated now and it's confusing that some places still use column_family whereas others table. Look at replica/database.hh and you'll see classes like column_family_stats and table_stats, and both being used by replica::table :-(

denesb commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2328/

There are some tests failing, please look into it.

raphaelsc commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2328/
> 
> There are some tests failing, please look into it.

Will fix and resend. thanks

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2357/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2358/

denesb commented:
Where changelog?

raphaelsc commented:
> Where changelog?

v2:
- rebase
- added [sstable_resharding_test: Switch to table_for_tests](https://github.com/scylladb/scylladb/pull/11592/commits/1e5e4c46365407be56ca7bb307a0d1e05694b735)
- included path to rename column_family_for_tests to table_for_tests
- fixed failing tests by using stable pointer (table_for_tests::data) instead in table_state.

avikivity commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -210,18 +210,10 @@ SEASTAR_TEST_CASE(compaction_manager_basic_test) {
-    cf->mark_ready_for_writes();
+    column_family_for_tests cf(env.manager(), s, tmp.path().string());
+    auto& cm = cf.get_compaction_manager();
+    auto close_cf = deferred_stop(cf);
```
With table_state, we can decouple compaction tests from replica::table completely.

raphaelsc replied:
That's indeed the goal.

avikivity commented:
Looks good, I left a couple of questions.

What about backlog computations? I think they need to be done per compaction group now (and summed to get the overall backlog), but the memory input needs to be divided by the number of compaction groups to account for the fact that each LSM tree is smaller by that factor.

raphaelsc commented:
> Looks good, I left a couple of questions.
> 
> What about backlog computations? I think they need to be done per compaction group now (and summed to get the overall backlog), but the memory input needs to be divided by the number of compaction groups to account for the fact that each LSM tree is smaller by that factor.

Agree. I said in the cover letter that backlog calculation per group will come next. The series is more than 10 patches already so I decided to split even the preliminary work. The next series should contain the backlog work in addition to introducing static groups via config option.

avikivity commented:
Where's the extracted fix? I assume we should merge it first?

raphaelsc commented:
> Where's the extracted fix? I assume we should merge it first?

Oops, missed ur question. Yes, it should be merged first. Extracting it now and will open a GH issue too.

raphaelsc commented:
> Looks good, I left a couple of questions.
> 
> What about backlog computations? I think they need to be done per compaction group now (and summed to get the overall backlog), but the memory input needs to be divided by the number of compaction groups to account for the fact that each LSM tree is smaller by that factor.

With a single group, a table produces backlog B 

If a table has N groups, each group's LSM tree will be smaller by N (with efficient partitioning, I'd expect that L0 sstables will be smaller by N as memtable size for each group will be smaller by N too).

So the backlog for each group is B / N 

But when summing the backlog of all groups, we have B / N * N = B.

The normalization happens on the sum of backlog of all trackers, so I don't think we have to account for number of groups when normalizing the backlog.

But, my intention (as documented in the cover letter) is to decouple the backlog tracker from compaction_strategy, allowing a given table to have multiple trackers (one for each group)

raphaelsc commented:
> > Where's the extracted fix? I assume we should merge it first?
> 
> Oops, missed ur question. Yes, it should be merged first. Extracting it now and will open a GH issue too.

Done: https://github.com/scylladb/scylladb/pull/11682

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2570/

raphaelsc commented:
v3: rebased

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2571/

raphaelsc commented:
Test Result (2 failures / ±0)
test_topology.test_remove_node_add_column
non-boost tests.test_topology.release.2415

@kbr- does it ring a bell to you?

kbr- commented:
We suspect this PR will fix it: https://github.com/scylladb/scylladb/pull/11691
It's currently in review.

raphaelsc commented:
> We suspect this PR will fix it: #11691 It's currently in review.

thanks.
